### PR TITLE
Ft/topbar insert menu

### PR DESCRIPTION
**Pull request by linelson**
Plugin for an Insert menu that provides GUI forms to edit v3 spec

### Description

This PR adds an insert menu plugin for OAS v3 documents and adds it to the standalone "Topbar" plugin.  

The insert menu provides options to add to the document via a form UI in a modal for adding paths, operations, info, external documentation, tag declarations, tags to an operation, and servers. The TopbarInsert component renders the dropdown insert menu (See topbar-insert.jsx), and the AddForm component (/forms/components/AddForm.jsx) renders the form. 

The plugin describes each form as an object that defines the form's behavior and how the form UI will render. See src/standalone/topbar-insert/forms/form-data-readme.md for more details.

Folder structure:
```
  /topbar-insert
    /dropdown    
    /forms
      /components
        AddForm.jsx                   Component to render the form.
        FormChild.jsx                  Component to render a nested child form.
        FormDropdown.jsx         Component that renders a dropdown form.
        FormInput.jsx                  Component that renders a basic form input.
        FormInputWrapper.jsx    Component that renders the name + description for a form field.
        FormMap.jsx                   Component that renders a form for a (key, value) pair
        forms.less                        Styling for the above components.
      /formObjects                     Contains objects describing forms.
      /helpers
        form-data-helpers.js     Recursively generates the form UI for a form object
        validation-helpers.js     Handles validation for a form object.
    form-data-readme.md         
    index.js                             Entry point for the plugin.
    topbar-insert.jsx               Top-level component that renders the insert menu.
```

### Motivation and Context

Even for users familiar with the Open API v3 specification, writing an open api document from scratch can be difficult and requires looking up definitions in the spec documentation. The intent of this feature is to make the creation and editing of a document simpler by providing a form UI that allows users to add to their document without having to worry as much about the rules of the spec. 

### How Has This Been Tested?

The feature has been tested manually by repeatedly using the every option in the insert menu to add to several different v3 documents while running the code locally using the dev server in both Edge (Windows 10 v42.17134.1.0) and Chrome (Windows 10 v.69.0.3497.100).  

Tests for this feature are in the /test/unit/standalone/topbar-insert folder and should run when ```npm run test``` is run. 

### Screenshots (if appropriate):
image
image
image
image
image
image
image
image

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [x] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by shockey** [comment]
please build

**Comment by JonathanParrilla** [comment]
please build

**Comment by shockey** [comment]
please build

**Comment by shockey** [comment]
please build

**Comment by shockey** [comment]
please build

**Comment by shockey** [comment]
please build

File: src/standalone/topbar-insert/forms/helpers/form-data-helpers.js [hunk 1]
```diff
@@ -0,0 +1,232 @@
+/* eslint react/jsx-no-bind: 0 */
+/* eslint no-use-before-define: 0 */
+import { OrderedMap, Map, List } from "immutable"
+import React from "react"
+import { checkForEmptyValue } from "./validation-helpers"
+import FormInputWrapper from "../components/FormInputWrapper"
```
[the remark this answers is not in the document]
Comment (shockey) [reply src/standalone/topbar-insert/forms/helpers/form-data-helpers.js]: Backlogging - this is housekeeping that can be done this week.

File: webpack-standalone.config.js [hunk 2]
```diff
@@ -10,13 +10,14 @@ module.exports = require("./make-webpack-config.js")({
...
     "swagger-editor-standalone-preset": [
+      "./src/standalone/styles/main.less",
       "./src/polyfills.js",
       "./src/standalone/index.js"
     ]
   },
 
   externals : {
-    react: 'React'
+    react: "react"
```
[the remark this answers is not in the document]
Comment (shockey) [reply webpack-standalone.config.js]: This ended up breaking our production build - I reverted it in https://github.com/swagger-api/swagger-editor/pull/1902.

@linelson, let me know if there's a pressing reason to reintroduce this!

[the remark this answers is not in the document]
Comment (linelson) [reply webpack-standalone.config.js]: The reason for the change was that the casing ```React``` was breaking webpack-dev-server for a React app created using create-react-app https://github.com/facebook/create-react-app when the standalone preset was imported from https://www.npmjs.com/package/swagger-editor-dist. 

Here's the error:
```
Failed to compile.
./node_modules/swagger-editor-dist/swagger-editor-standalone-preset.js
Module not found: `C:\Users\linels\repos\react-app\src\node_modules\React\index.js` does not match the corresponding path on disk `react`.
```
[CaseSensitivePathsPlugin](https://www.npmjs.com/package/case-sensitive-paths-webpack-plugin), which is used by create-react-app, appears to be the culprit.
[Ejecting](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#npm-run-eject) create-react-app and removing dependencies on CaseSensitivePathsPlugin allows webpack-dev-server to run without problems with either casing.

When the standalone preset bundle is created with casing ```react: 'react'``` webpack-dev-server works as expected without needing to ```eject``` and remove dependencies on CaseSensitivePathsPlugin.

File: src/standalone/topbar-insert/forms/components/FormChild.jsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import React from "react"
+import PropTypes from "prop-types"
+import FormInputWrapper from "./FormInputWrapper"
+
+const FormChild = (props) => (
+  <div key={props.name} className="card-body">
+    <FormInputWrapper name={props.name} description={props.description} isRequired={props.isRequired}>
+        { !props.isRequired && <a href="javascript:void(0)" onClick={props.flipRequired}> Add {props.name} </a> }
```
[the remark this answers is not in the document]
Comment (linelson) [reply src/standalone/topbar-insert/forms/components/FormChild.jsx]: No objections here to its removal.

File: src/standalone/topbar-insert/forms/helpers/form-data-helpers.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (linelson) [reply src/standalone/topbar-insert/forms/helpers/form-data-helpers.js]: Let me know if you'd like me to send a PR for this.

**Comment by linelson** [comment]
@shockey thanks for getting this merged in!

File: src/standalone/topbar-insert/forms/helpers/form-data-helpers.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (shockey) [reply src/standalone/topbar-insert/forms/helpers/form-data-helpers.js]: @linelson, if you have the time that'd be wonderful 😄

**Pull request merged**